src/DssProxyActionsCropper.sol

```diff
@@ -0,0 +1,815 @@
+        require(CdpRegistryLike(cdpRegistry).ilks(cdp) == GemJoinLike(gemJoin).ilk(), "wrong-ilk");
+
+        // Exits token amount to proxy address as a token
+        CropperLike(cropper).exit(gemJoin, address(this), amt);
+        // Exits token amount to the user's wallet as a token
+        GemJoinLike(gemJoin).gem().transfer(msg.sender, amt);
+    }
+
+    function fleeETH(
+        address ethJoin,
+        uint256 cdp
+    ) external {
+        require(CdpRegistryLike(cdpRegistry).owns(cdp) == address(this), "wrong-cdp");
+        require(CdpRegistryLike(cdpRegistry).ilks(cdp) == GemJoinLike(ethJoin).ilk(), "wrong-ilk");
+        GemLike gem = GemJoinLike(ethJoin).gem();
+
+        // Exits WETH to proxy address as a token
+        uint256 pre = gem.balanceOf(address(this));
+        CropperLike(cropper).flee(ethJoin, address(this));
+        uint256 wad = _sub(gem.balanceOf(address(this)), pre);
+
+        // Converts WETH to ETH
+        gem.withdraw(wad);
+        // Sends ETH back to the user's wallet
+        msg.sender.transfer(wad);
+    }
+
+    function fleeGem(
+        address gemJoin,
+        uint256 cdp
+    ) external {
+        require(CdpRegistryLike(cdpRegistry).owns(cdp) == address(this), "wrong-cdp");
+        require(CdpRegistryLike(cdpRegistry).ilks(cdp) == GemJoinLike(gemJoin).ilk(), "wrong-ilk");
+
+        // Exits token amount to the user's wallet as a token
+        CropperLike(cropper).flee(gemJoin, msg.sender);
+    }
+
+    function draw(
+        address jug,
+        address daiJoin,
+        uint256 cdp,
+        uint256 wad
+    ) external {
+        address owner = CdpRegistryLike(cdpRegistry).owns(cdp);
+        bytes32 ilk = CdpRegistryLike(cdpRegistry).ilks(cdp);
+
+        // Generates debt in the CDP
+        _frob(ilk, owner, 0, _getDrawDart(jug, owner, ilk, wad));
+        // Allows adapter to access to proxy's DAI balance in the vat
+        if (VatLike(vat).can(address(this), address(daiJoin)) == 0) {
+            VatLike(vat).hope(daiJoin);
+        }
+        // Exits DAI to the user's wallet as a token
+        DaiJoinLike(daiJoin).exit(msg.sender, wad);
+    }
+
+    function wipe(
+        address daiJoin,
+        uint256 cdp,
+        uint256 wad
+    ) external {
+        address owner = CdpRegistryLike(cdpRegistry).owns(cdp);
+        bytes32 ilk = CdpRegistryLike(cdpRegistry).ilks(cdp);
+
+        // Joins DAI amount into the vat
+        daiJoin_join(daiJoin, owner, wad);
+        // Allows cropper to access to proxy's DAI balance in the vat
+        VatLike(vat).hope(cropper);
+        // Paybacks debt to the CDP
+        _frob(
+            ilk,
+            owner,
+            0,
+            _getWipeDart(
+                VatLike(vat).dai(owner),
+                owner,
+                ilk
+            )
+        );
+        // Denies cropper to access to proxy's DAI balance in the vat after execution
+        VatLike(vat).nope(cropper);
```

Comment:
I know we did it for charter as well, but what was the reason of denying after?
Proxy's vat dai funds should still be safe even leaving the `hope` active, right?

Reply:
We said (sorry for quoting, just tryong to bring it accurately) -

"the hope/nope thing might be an extra safety measure before we can analyse better the potential of leaving a long term DAI allowance to the charter. I think today won't be a problem as the frob function is quite restrictive but might be more flexible in the future. So I think is better just to turn off the allowance at end."

I don't see a problem with removing it for this version as we don't rely on allowance but on access control through `allowed(x)` and `msg.sender`. Nevertheless as charter and cropper are upgradeable such remaining approvals can be dangerous in the future, so I suggest to leave it, at least for this version.

Reply:
Yeah looks goos, let's leave it, just wanted to make sure I was not missing anything.